File: javascript/channel_status.js
#!/usr/bin/env node
// channel_status.pl — portiert nach javascript
// Quelle: perl5, Projects@abstractions:perl5/channel_status.pl
// Erzeugt: 2026-08-08 durch ABSTRACTIONS_MANAGER.py

// channel_status.js — portiert nach JavaScript für Node 20
// Quelle: perl5, channel_status.pl
// Erzeugt: 2026-08-07 durch ABSTRACTIONS_MANAGER.py

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const { Command } = require('commander');

// Konfiguration
const WORKSPACE = "/home/openclaw/.openclaw/workspace";
const LOGS_DB = `${WORKSPACE}/db/logs.db`;
const CONFIG_FILE = `${WORKSPACE}/config/channel-status.json`;
const LOG_FILE = `${WORKSPACE}/logs/channel-status.log`;

function logMessage(message, level = "INFO") {
    const timestamp = new Date().toISOString().replace('T', ' ').substring(0, 19);
    const entry = `[${timestamp}] [${level}] ${message}\n`;
    process.stdout.write(entry);
    fs.appendFileSync(LOG_FILE, entry);
}

function getSystemStatus() {
    const status = {
        timestamp: new Date().toISOString().replace('T', ' ').substring(0, 19),
        nodes: {},
        agents: {},
        system: {}
    };
    
    // Node-Status (vereinfacht)
    const nodes = {
        node1: {name: "Gateway", status: "online"},
        node2: {name: "Worker", status: "online"},
        node3: {name: "Relay", status: "offline", reason: "disk full"},
        node5: {name: "Redmi", status: "intermittent"},
        node7: {name: "Docker", status: "planned"}
    };
    status.nodes = nodes;
    
    // Agent-Status aus Cron
    try {
        const stdout = execSync('crontab -l', { encoding: 'utf8' });
        const lines = stdout.split('\n');
        const cronLines = lines.filter(line => !/^\s*#/.test(line) && /\S/.test(line)).length;
        status.agents.active_crons = cronLines;
    } catch (error) {
        status.agents.active_crons = "unknown";
    }
    
    // System-Metriken
    try {
        // Disk usage
        const dfOutput = execSync('df -h /', { encoding: 'utf8' });
        const dfLines = dfOutput.split('\n');
        for (const line of dfLines) {
            if (line.includes('/') && line.includes('%')) {
                const parts = line.split(/\s+/);
                status.system.disk_used = parts[4];
                break;
            }
        }
        
        // RAM usage
        const freeOutput = execSync('free -h', { encoding: 'utf8' });
        const freeLines = freeOutput.split('\n');
        for (const line of freeLines) {
            if (line.includes('Mem:')) {
                const parts = line.split(/\s+/);
                status.system.ram_total = parts[1];
                status.system.ram_used = parts[2];
                break;
            }
        }
    } catch (error) {
        // Fehler werden ignoriert, status.system bleibt leer
    }
    
    return status;
}

function formatDailyStatus(status) {
    const nodes = status.nodes;
    let online = 0;
    for (const nodeId in nodes) {
        if (nodes[nodeId].status === "online") {
            online++;
        }
    }
    
    let message = "📊 **Täglicher Status-Report**\n";
    message += "🗓️ " + new Date().toISOString().replace('T', ' ').substring(0, 16) + "\n\n";
    message += `**🖥️ Nodes (${online}/5 online):**\n`;
    
    for (const nodeId of Object.keys(nodes).sort()) {
        const info = nodes[nodeId];
        const emoji = info.status === "online" ? "🟢" : 
                      (info.status === "offline" ? "🔴" : "🟡");
        message += `${emoji} ${info.name}: ${info.status}`;
        if (info.reason) {
            message += ` (${info.reason})`;
        }
        message += "\n";
    }
    
    message += "\n**🤖 Agents:**\n";
    message += `Aktive Cron-Jobs: ${status.agents.active_crons}\n`;
    
    if (status.system.disk_used) {
        message += "\n**💾 System:**\n";
        message += `Disk: ${status.system.disk_used} belegt\n`;
        message += `RAM: ${status.system.ram_used} / ${status.system.ram_total}\n`;
    }
    
    return message;
}

function formatWeeklyStatus(status) {
    const now = new Date();
    const weekNumber = Math.ceil((((now - new Date(now.getFullYear(), 0, 1)) / 86400000) + now.getDay() + 1) / 7);
    let message = "📈 **Wöchentlicher Report**\n";
    message += `📅 Woche ${weekNumber} - ${now.getFullYear()}\n\n`;
    message += "**Zusammenfassung:**\n";
    message += "- 5 aktive Sub-Agents\n";
    message += "- 11 Skills synchronisiert\n";
    message += "- 3 neue Features implementiert\n\n";
    message += "**Top-Ereignisse:**\n";
    message += "1. ClawHub-Git Sync implementiert ✅\n";
    message += "2. Node 3 Disk voll (95%) ⚠️\n";
    message += "3. Channel-Status-Agent aktiviert 🆕\n\n";
    message += "**Geplante Wartungen:**\n";
    message += "- Node 3: Disk-Cleanup erforderlich\n";
    message += "- Node 7: Docker-Setup ausstehend\n";
    return message;
}

function sendToChannel(message, channelType = "telegram", channelId = "-1002381931352") {
    if (channelType === "telegram") {
        // Nutze OpenClaw message tool
        const cmd = `openclaw message send --target ${channelId} --message "${message.replace(/"/g, '\\"')}"`;
        try {
            execSync(cmd, { stdio: 'pipe' });
            logMessage(`Message sent to ${channelType} ${channelId}`);
            return true;
        } catch (error) {
            logMessage(`Failed to send: ${error.message}`, "ERROR");
            return false;
        }
    } else {
        logMessage(`Channel type ${channelType} not implemented`, "WARN");
        return false;
    }
}

function main() {
    const program = new Command();
    
    program
        .option('--type <type>', 'Type of status update (daily|weekly|alert)')
        .option('--message <message>', 'Message content for alert type')
        .option('--channel <channel>', 'Channel ID', '-1002381931352')
        .option('--dry-run', 'Dry run mode')
        .parse();
    
    const options = program.opts();
    
    if (!options.type) {
        console.error("Type is required\n");
        process.exit(1);
    }
    
    if (!['daily', 'weekly', 'alert'].includes(options.type)) {
        console.error(`Invalid type: ${options.type}\n`);
        process.exit(1);
    }
    
    logMessage(`Starting ${options.type} status update`);
    
    // Status sammeln
    const status = getSystemStatus();
    
    // Message formatieren
    let formattedMessage;
    if (options.type === 'daily') {
        formattedMessage = formatDailyStatus(status);
    } else if (options.type === 'weekly') {
        formattedMessage = formatWeeklyStatus(status);
    } else if (options.type === 'alert') {
        formattedMessage = "🚨 **ALERT**\n" + (options.message || 'Manual alert');
    }
    
    // Senden oder Dry-Run
    if (options.dryRun) {
        process.stdout.write("\n--- DRY RUN ---\n");
        process.stdout.write(formattedMessage);
        process.stdout.write("\n--- END ---\n");
    } else {
        sendToChannel(formattedMessage, "telegram", options.channel);
    }
    
    logMessage("Status update completed");
}

// Ensure log directory exists
const logDir = path.dirname(LOG_FILE);
if (!fs.existsSync(logDir)) {
    fs.mkdirSync(logDir, { recursive: true });
}

main();

File: python/channel_status.py
#!/usr/bin/env python3
# channel_status.js — portiert nach python
# Quelle: javascript, Projects@abstractions:javascript/channel_status.js
# Erzeugt: 2026-08-08 durch ABSTRACTIONS_MANAGER.py

# channel_status.py — portiert nach javascript
# Quelle: python, OpenClaw@gateway1:skills/channel-status-agent/scripts/channel_status.py
# auch in: OpenClaw@gateway2:skills/channel-status-agent/scripts/channel_status.py
# Erzeugt: 2026-08-07 durch ABSTRACTIONS_MANAGER.py

"""
Channel Status Agent - Automatische Status-Updates
"""

import os
import subprocess
import json
import argparse
from datetime import datetime
from pathlib import Path

# Konfiguration
WORKSPACE = Path('/home/openclaw/.openclaw/workspace')
LOGS_DB = WORKSPACE / 'db/logs.db'
CONFIG_FILE = WORKSPACE / 'config/channel-status.json'
LOG_FILE = WORKSPACE / 'logs/channel-status.log'

def log(message, level="INFO"):
    """Logging"""
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    entry = f"[{timestamp}] [{level}] {message}"
    print(entry)
    with open(LOG_FILE, 'a') as f:
        f.write(entry + '\n')

def get_system_status():
    """Sammelt System-Status"""
    status = {
        "timestamp": datetime.now().isoformat(),
        "nodes": {},
        "agents": {},
        "system": {}
    }
    
    # Node-Status (vereinfacht)
    nodes = {
        "node1": {"name": "Gateway", "status": "online"},
        "node2": {"name": "Worker", "status": "online"},
        "node3": {"name": "Relay", "status": "offline", "reason": "disk full"},
        "node5": {"name": "Redmi", "status": "intermittent"},
        "node7": {"name": "Docker", "status": "planned"}
    }
    status["nodes"] = nodes
    
    # Agent-Status aus Cron
    try:
        result = subprocess.run(['crontab', '-l'], capture_output=True, text=True, check=True)
        cron_lines = [line for line in result.stdout.split('\n') if line and not line.startswith('#')]
        status["agents"]["active_crons"] = len(cron_lines)
    except subprocess.CalledProcessError:
        status["agents"]["active_crons"] = "unknown"
    
    # System-Metriken
    try:
        # Disk usage
        df_result = subprocess.run(['df', '-h', '/'], capture_output=True, text=True, check=True)
        for line in df_result.stdout.split('\n'):
            if '/' in line and '%' in line:
                parts = line.strip().split()
                status["system"]["disk_used"] = parts[4]
                break
        
        # RAM usage
        free_result = subprocess.run(['free', '-h'], capture_output=True, text=True, check=True)
        for line in free_result.stdout.split('\n'):
            if 'Mem:' in line:
                parts = line.strip().split()
                status["system"]["ram_total"] = parts[1]
                status["system"]["ram_used"] = parts[2]
                break
    except subprocess.CalledProcessError:
        # Ignore errors
        pass
    
    return status

def format_daily_status(status):
    """Formatiert täglichen Status"""
    nodes = status["nodes"]
    online = len([n for n in nodes.values() if n["status"] == "online"])
    
    message = f"""📊 **Täglicher Status-Report**
🗓️ {datetime.now().strftime('%d.%m.%Y %H:%M')}

**🖥️ Nodes ({online}/5 online):**
"""
    
    for node_id, info in nodes.items():
        emoji = "🟢" if info["status"] == "online" else "🔴" if info["status"] == "offline" else "🟡"
        message += f"{emoji} {info['name']}: {info['status']}"
        if "reason" in info:
            message += f" ({info['reason']})"
        message += "\n"
    
    message += f"\n**🤖 Agents:**\n"
    message += f"Aktive Cron-Jobs: {status['agents']['active_crons']}\n"
    
    if "disk_used" in status["system"]:
        message += f"\n**💾 System:**\n"
        message += f"Disk: {status['system']['disk_used']} belegt\n"
        message += f"RAM: {status['system']['ram_used']} / {status['system']['ram_total']}\n"
    
    return message

def format_weekly_status(status):
    """Formatiert wöchentlichen Status"""
    now = datetime.now()
    week_number = now.isocalendar()[1]
    
    return f"""📈 **Wöchentlicher Report**
📅 Woche {week_number:02d} - {now.year}

**Zusammenfassung:**
- 5 aktive Sub-Agents
- 11 Skills synchronisiert
- 3 neue Features implementiert

**Top-Ereignisse:**
1. ClawHub-Git Sync implementiert ✅
2. Node 3 Disk voll (95%) ⚠️
3. Channel-Status-Agent aktiviert 🆕

**Geplante Wartungen:**
- Node 3: Disk-Cleanup erforderlich
- Node 7: Docker-Setup ausstehend
"""

def send_to_channel(message, channel_type="telegram", channel_id="-1002381931352"):
    """Sendet Nachricht an Channel"""
    if channel_type == "telegram":
        # Nutze OpenClaw message tool
        cmd = ['openclaw', 'message', 'send', '--target', channel_id, '--message', message]
    else:
        log(f"Channel type {channel_type} not implemented", "WARN")
        return False
    
    try:
        subprocess.run(cmd, check=True, capture_output=True)
        log(f"Message sent to {channel_type} {channel_id}")
        return True
    except subprocess.CalledProcessError as e:
        log(f"Failed to send: {e}", "ERROR")
        return False

def main():
    """Hauptfunktion"""
    parser = argparse.ArgumentParser(description='Channel Status Agent')
    parser.add_argument('--type', choices=['daily', 'weekly', 'alert'], required=True,
                        help='Type of status update')
    parser.add_argument('--message', type=str, help='Alert message')
    parser.add_argument('--channel', type=str, default='-1002381931352',
                        help='Channel ID')
    parser.add_argument('--dry-run', action='store_true',
                        help='Show message without sending')
    
    args = parser.parse_args()
    
    log(f"Starting {args.type} status update")
    
    # Status sammeln
    status = get_system_status()
    
    # Message formatieren
    if args.type == 'daily':
        message = format_daily_status(status)
    elif args.type == 'weekly':
        message = format_weekly_status(status)
    elif args.type == 'alert':
        message = f"🚨 **ALERT**\n{args.message or 'Manual alert'}"
    
    # Senden oder Dry-Run
    if args.dry_run:
        print("\n--- DRY RUN ---")
        print(message)
        print("--- END ---")
    else:
        send_to_channel(message, "telegram", args.channel)
    
    log("Status update completed")

# Ensure log directory exists
LOG_FILE.parent.mkdir(parents=True, exist_ok=True)

if __name__ == "__main__":
    main()
